File: src/domain/display_names.py
from __future__ import annotations


_EMPTY_NAME_MARKERS = frozenset({"", "none", "null", "undefined"})


def normalize_display_text(value: object) -> str | None:
    if value is None:
        return None
    normalized = str(value).strip()
    if normalized.lower() in _EMPTY_NAME_MARKERS:
        return None
    return normalized


def join_name_parts(
    first_name: object | None,
    last_name: object | None,
) -> str | None:
    parts = [
        part
        for part in (
            normalize_display_text(first_name),
            normalize_display_text(last_name),
        )
        if part
    ]
    if not parts:
        return None
    return " ".join(parts)


def normalize_username(username: object | None) -> str | None:
    normalized = normalize_display_text(username)
    if not normalized:
        return None
    return normalized if normalized.startswith("@") else f"@{normalized}"


def build_display_name(
    *,
    full_name: object | None = None,
    first_name: object | None = None,
    last_name: object | None = None,
    username: object | None = None,
    email: object | None = None,
    fallback: str,
) -> str:
    return (
        normalize_display_text(full_name)
        or join_name_parts(first_name, last_name)
        or normalize_username(username)
        or normalize_display_text(email)
        or fallback
    )

File: frontend/src/shared/lib/displayNames.ts
export type DisplayNameLike = {
  display_name?: string | null;
  full_name?: string | null;
  first_name?: string | null;
  last_name?: string | null;
  username?: string | null;
  email?: string | null;
};

const normalizeText = (value?: string | null): string | null => {
  const normalized = value?.trim();
  if (!normalized) return null;

  const lowered = normalized.toLowerCase();
  if (lowered === 'none' || lowered === 'null' || lowered === 'undefined') {
    return null;
  }

  return normalized;
};

const normalizeUsername = (username?: string | null): string | null => {
  const value = normalizeText(username);
  if (!value) return null;
  return value.startsWith('@') ? value : `@${value}`;
};

const joinNameParts = (
  firstName?: string | null,
  lastName?: string | null,
): string | null => {
  const parts = [normalizeText(firstName), normalizeText(lastName)].filter(
    (part): part is string => Boolean(part),
  );

  return parts.length ? parts.join(' ') : null;
};

export const getDisplayName = (
  entity?: DisplayNameLike | null,
  fallback = 'Клиент',
): string => {
  return (
    normalizeText(entity?.display_name) ||
    normalizeText(entity?.full_name) ||
    joinNameParts(entity?.first_name, entity?.last_name) ||
    normalizeUsername(entity?.username) ||
    normalizeText(entity?.email) ||
    fallback
  );
};

export const getSecondaryDisplayText = (
  entity?: DisplayNameLike | null,
): string | null => {
  const primary = getDisplayName(entity, '');
  const username = normalizeUsername(entity?.username);
  const fullName = normalizeText(entity?.full_name);
  const email = normalizeText(entity?.email);

  if (primary && username && primary !== username) return username;
  if (primary && fullName && primary !== fullName) return fullName;
  if (primary && email && primary !== email) return email;
  return null;
};

export const getDisplayInitials = (
  entity?: DisplayNameLike | null,
  fallback = 'Клиент',
): string => {
  const label = getDisplayName(entity, fallback).replace(/^@/, '');
  const parts = label.split(/\s+/).filter(Boolean);

  if (parts.length === 0) return fallback.slice(0, 2).toUpperCase();
  if (parts.length === 1) return parts[0].slice(0, 2).toUpperCase();

  return `${parts[0][0] ?? ''}${parts[1][0] ?? ''}`.toUpperCase();
};
